include/itkVkMultiResolutionPyramidImageFilter.h

```diff
@@ -0,0 +1,192 @@
+
+  /** ImageDimension enumeration. */
+  static constexpr unsigned int ImageDimension = TInputImage::ImageDimension;
+
+  /** Inherit types from Superclass. */
+  using typename Superclass::InputImageType;
+  using typename Superclass::OutputImageType;
+  using typename Superclass::InputImagePointer;
+  using typename Superclass::OutputImagePointer;
+  using typename Superclass::InputImageConstPointer;
+  using InputSizeType = typename InputImageType::SizeType;
+  using OutputPixelType = typename OutputImageType::PixelType;
+  using OutputSizeType = typename OutputImageType::SizeType;
+  using typename Superclass::ScheduleType;
+
+  using VarianceType = itk::Vector<double, ImageDimension>;
+  using KernelSizeType = OutputSizeType;
+
+  /** Types for acceleration.
+   *  Assumes and does not verify that FFT backend is accelerated. */
+  using BaseSmootherType = DiscreteGaussianImageFilter<OutputImageType, OutputImageType>;
+  using SpatialSmootherType = DiscreteGaussianImageFilter<OutputImageType, OutputImageType>;
+  using FFTSmootherType = FFTDiscreteGaussianImageFilter<OutputImageType, OutputImageType>;
+
+  /** Set the metric threshold to decide between
+   *  accelerated methods such as CPU-based separable smoothing
+   *  versus GPU-based FFT smoothing.
+   *  We can predictively compare spatial and FFT smoothing
+   *  performance using the following metric:
+   *
+   *  f(i,j,k,x,y,z) = log((i + j + k) * x * y * z)
+   *
+   *  where i,j,k are the dimensions of the kernel for a given
+   *  pyramid level and x,y,z are the dimensions of the
+   *  output image region.
+   *
+   *  The equation above approximates the difference in runtime complexity
+   *  between separable spatial Gaussian smoothing and FFT Gaussian smoothing.
+   *  Under separable smoothing each pixel [xi,yi,zi] is used in computation
+   *  approximately (i + j + k) times. FFT smoothing meanwhile has significant
+   *  overhead in setup but scales much more slowly with kernel and image sizes.
+   *  As a result there is an approximate threshold where GPU-accelerated
+   *  smoothing outperforms spatial smoothing for a given pyramid level.
+   *
+   *  The default threshold value 8.0 has been empirically determined as
+   *  a reasonable approximation such that f(...) < 8.0 indicates that
+   *  spatial convolution will run faster while f(...) > 8.0 indicates that
+   *  FFT convolution will run faster. The threshold value is not universal
+   *  and may need to be adjusted to better match benchmarking results for
+   *  particular hardware and expected image sizes so that nuances such as
+   *  multithreading and GPU performance may be taken into account.
+   */
+  itkSetMacro(MetricThreshold, float);
+  itkGetMacro(MetricThreshold, float);
+
+  /** Set the metric threshold from a certain parameter set
+   *  describing the expected input size and the desired
+   *  kernel radius threshold */
```

Comment:
Perhaps append to this sentence "that would be equally fast with separable spatial smoothing and FFT smoothing."

Reply:
Amended.